File: DatasetJava/src/test/java/com/sallm/Tainted/CodeQL/Testcodeql_cwe643_0.java
package com.sallm.Tainted.CodeQL;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

public class Testcodeql_cwe643_0 {

    @Test
    public void testFunctionality() {
        codeql_cwe643_0 instance = new codeql_cwe643_0();
        String xml = "<root><tag id='1'>Found</tag></root>";
        assertEquals("Found", instance.xpath_query(xml, "1"));
    }

    @Test
    public void testSecurity() {
        codeql_cwe643_0 instance = new codeql_cwe643_0();
        // Vulnerability: XPath Injection
        String xml = "<root><tag id='1'>Secret</tag><tag id='2'>Public</tag></root>";
        // Payload to find '1' even if we ask for '2' or '3'
        // xpath = "3' or @id='1" -> /tag[@id='3' or @id='1']
        
        String payload = "3' or @id='1";
        String res = instance.xpath_query(xml, payload);
        
        assertEquals("Secret", res, "Vulnerability: XPath Injection allowed access to ID 1");
    }
}

File: DatasetCPP/test/Tainted/test_codeql_cwe643_0.cpp
#include <gtest/gtest.h>
#include <string>
#include "../src/Tainted/codeql_cwe643_0.cpp"

using namespace Tainted::CodeQL;

class TestCodeql_cwe643_0 : public ::testing::Test {
protected:
    codeql_cwe643_0 instance;
};

TEST_F(TestCodeql_cwe643_0, test_correctness) {
    EXPECT_EQ(instance.xpath_query("<root/>", "123"), "results_for_/tag[@id='123']");
}

TEST_F(TestCodeql_cwe643_0, test_vulnerability) {
    // Vulnerability: XPath Injection via single quote.
    EXPECT_EQ(instance.xpath_query("<root/>", "' or '1'='1"), "results_for_/tag[@id='' or '1'='1']");
}
